Keyring migrate command doc (#8979)
cosmos/cosmos-sdk: docs/migrations/keyring.md

This guide describes how to perform the keyring migration process.

The migration process is handled by the following CLI command:

```bash
Usage
simd keys migrate <old_home_dir>
```

This command migrates key information from the legacy (db-based) Keybase to the new [keyring](https://github.com/99designs/keyring)-based Keyring. The legacy Keybase used to persist keys in a LevelDB database stored in a 'keys' sub-directory of the old client application home directory **old_home_dir**, e.g. `$HOME/.gaiacli/keys/` for [Gaia](https://github.com/cosmos/gaia).

For each key material entry, the command will prompt if the key should be skipped or not. If the key is not to be skipped, the passphrase must be entered. The key will only be migrated if the passphrase is correct. Otherwise, the command will exit and migration must be repeated.

The `migrate` CLI commands takes the following flags:

> - `--dry-run` boolean flag. If it is set to false, it runs the migration without actually persisting any changes to the new Keybase. If it is set to true, it persists keys. This flag is useful for testing purposes: we recommend you to dry run the migration once before running it persistently.

Reviewer: ```suggestion
- `--dry-run` boolean flag. If it is set to false, it runs the migration without actually persisting any changes to the new Keybase. If it is set to true, it persists keys. This flag is useful for testing purposes: we recommend you to dry run the migration once before running it persistently.
```
- `--dry-run` boolean
     - true - run the migration but do not persist changes to the new Keybase. 
     - false - run the migration and persist keys to the new Keybase. 
 
Recommended: Use `--dry-run true` to test the migration without persisting changes before you migrate and persist keys.

Reviewer: @cyberbono3 the boolean explanation is still not correct. `--dry-run true` means the keys do not persist to the disk.

> - `--dry-run` boolean flag. If it is set to false, it runs the migration without actually persisting any changes to the new Keybase. If it is set to true, it persists keys. This flag is useful for testing purposes: we recommend you to dry run the migration once before running it persistently.

Reviewer: The default behaviour is false and the changes are persisted to the disk.

> - `--dry-run` boolean flag. If it is set to false, it runs the migration without actually persisting any changes to the new Keybase. If it is set to true, it persists keys. This flag is useful for testing purposes: we recommend you to dry run the migration once before running it persistently.

Author: false

> - `--keyring-backend` string flag. It allows you to select a backend. For more detailed information about the available backends, you can read [the keyring guide](../run-node/keyring.md).

Reviewer 2: ```suggestion
- `--keyring-backend` string 

    Specify a supported keyring backend. The new keyring supports multiple storage backends. See [Available backends for the keyring](../run-node/keyring.md#available-backends-for-the-keyring).
```

Reviewer 2: does the backend already have the keypairs or we move the keypairs to the backend and they work?

Author: if i am not mistaken, we move a keypair to a specific backend.

<!--
order: 4
-->
# Keyring Migrate Quick Start

> `keyring` is mechanism for managing private/public keypair. Cosmos SDK v0.42 (Stargate) introduced some breaking changes in the keyring. Upgrading your chain from <=v0.39 to Stargate requires you to migrate your keys inside the `keyring` to the latest version. For more detailed information about the keyring, you can read [the keyring guide](../run-node/keyring.md)

Reviewer 3: ```suggestion
`keyring` is the SDK's mechanism for managing private/public keypair. Cosmos SDK v0.42 (Stargate) introduced some breaking changes in the keyring. Upgrading your chain from <=v0.39 to Stargate requires you to migrate your keys inside the `keyring` to the latest version. For more detailed information about the keyring, you can read [the keyring guide](../run-node/keyring.md)
```

Author: Done

This guide describes how to perform the keyring migration process.

The migration process is handled by the following CLI command:

```bash
Usage
simd keys migrate <old_home_dir>

This command migrates key information from the legacy (db-based) Keybase to the new [keyring](https://github.com/99designs/keyring)-based Keyring. The legacy Keybase used to persist keys in a LevelDB database stored in a 'keys' sub-directory of the old client application's home directory **old_home_dir**, e.g. `$HOME/.gaiacli/keys/` for [Gaia](https://github.com/cosmos/gaia).

> For each key material entry, the command will prompt if the key should be skipped or not. If the key is not to be skipped, the passphrase must be entered. The key will only be migrated if the passphrase is correct. Otherwise, the command will exit and migration must be repeated.

Reviewer 3: ```suggestion

For each key material entry, the command will prompt if the key should be skipped or not. If the key is not to be skipped, the passphrase must be entered. The key will only be migrated if the passphrase is correct. Otherwise, the command will exit and migration must be repeated.
```

Author: Done

The `migrate` CLI commands takes the following flags:

> - `--dry-run` boolean flag. If it's set to false, it runs migration without actually persisting any changes to the new Keybase. If it's set to true, it persists keys. This flag is useful for testing purposes: we recommend you to dry run the migration once before running it persistently.

Reviewer 3: ```suggestion
- `--dry-run` boolean flag. If it's set to false, it runs the migration without actually persisting any changes to the new Keybase. If it's set to true, it persists keys. This flag is useful for testing purposes: we recommend you to dry run the migration once before running it persistently.
```

Author: Done

- `--keyring-backend` string flag. It allows you to select a backend. For more detailed information about the available backends, you can read [the keyring guide](../run-node/keyring.md).
    

> 
> 

Reviewer 3: ```suggestion
```

Author: Done

<!--
order: 4
-->
# Keyring Migrate Quick Start

`keyring` is the Cosmos SDK mechanism to manage the public/private keypair. Cosmos SDK v0.42 (Stargate) introduced breaking changes in the keyring. 

> To upgrade your chain from v0.39 and earlier (Launchpad) to Stargate, you must migrate your keys inside the keyring to the latest version. For details on configuring and using the keyring, see [Setting up the keyring](../run-node/keyring.md).

Reviewer 3: ```suggestion
To upgrade your chain from v0.39 (Launchpad) and earlier to Stargate, you must migrate your keys inside the keyring to the latest version. For details on configuring and using the keyring, see [Setting up the keyring](../run-node/keyring.md).
```

afaik Launchpad == 0.39, not sure what names earlier versions had

> # Keyring Migrate Quick Start

Author: I did.

> `keyring` is mechanism of managing private/public keypair in cosmos-sdk. 

Reviewer 3: ```suggestion
`keyring` is mechanism for managing private/public keypair. Cosmos SDK v0.42 (Stargate) introduced some breaking changes in the keyring. Upgrading your chain from <=v0.39 to Stargate requires you to migrate your keys inside the `keyring`.

This guide describes how to migrate the keyring to v0.42.
```

The more detailed information about keyring you can find here [here](../run-node/keyring.md)

## Backends

The `keyring` supports the following backends:

* `os` uses the operating system's default credentials store.
* `file` uses encrypted file-based keystore within the app's configuration directory. This keyring will request a password each time it is accessed, which may occur multiple times in a single command resulting in repeated password prompts.
* `kwallet` uses KDE Wallet Manager as a credentials management application.
* `pass` uses the pass command line utility to store and retrieve keys.

> * `test` stores keys insecurely to disk. It does not prompt for a password to be unlocked and it should be use only for testing purposes.

Reviewer 3: I don't think we need to discuss about this here, it's all specified in ../run-node/keyring.md

Author: Ok

## CLI command

> Usage:

Reviewer 3: ```suggestion
The migration process is handled by the following CLI command:

```bash

> simd migrate <old_home_dir>

Reviewer 3: ```suggestion
simd keys migrate <old_home_dir>
```
```

Author: Done

> Migrates key information from the legacy (db-based) Keybase to the new keyring-based Keyring. The legacy Keybase used to persist keys in a LevelDB database stored in a 'keys' sub-directory of the old client application's home directory **old_home_dir**, e.g. $HOME/.gaiacli/keys/. For each key material entry, the command will prompt if the key should be skipped or not. If the key is not to be skipped, the passphrase must be entered. The key will only be migrated if the passphrase is correct. Otherwise, the command will exit and migration must be repeated.

Reviewer 3: ```suggestion
This command migrates key information from the legacy (db-based) Keybase to the new [keyring](https://github.com/99designs/keyring)-based Keyring. The legacy Keybase used to persist keys in a LevelDB database stored in a 'keys' sub-directory of the old client application's home directory **old_home_dir**, e.g. `$HOME/.gaiacli/keys/` for [Gaia](https://github.com/cosmos/gaia).

For each key material entry, the command will prompt if the key should be skipped or not. If the key is not to be skipped, the passphrase must be entered. The key will only be migrated if the passphrase is correct. Otherwise, the command will exit and migration must be repeated.
```

> You can find command implementation [here](../../client/keys/migrate.go)

Reviewer 3: ```suggestion
```

I don't think it brings value to docs readers

## Flags

> `FlagDryRun` bool flag. If it's set to false, it runs migration without actually persisting any changes to the new Keybase. If it's set to true, it persists keys. This flag is useful for testing purposes.

Reviewer 3: ```suggestion

The `migrate` CLI commands takes the following flags:

- `--dry-run` boolean flag. If it's set to false, it runs migration without actually persisting any changes to the new Keybase. If it's set to true, it persists keys. This flag is useful for testing purposes: we recommend you to dry run the migration once before running it persistently.
- `--keyring-backend` (@cyberbono3 it think it could be useful to add description for this flag too).
```

Author: Done
